File: java/src/main/java/br/com/emendes/removeelement/RemoveElement.java
package br.com.emendes.removeelement;

/**
 * Dado um array nums de inteiros e um inteiro val. remova todas as ocorrências de val em nums in-place.
 * A ordem relativa dos elementos pode ser alterada.<br><br>
 * <p>
 * Como é impossível alterar o comprimento do array em alguns idiomas,
 * você deve colocar o resultado na primeira parte do array nums.<br><br>
 * <p>
 * Mais formalmente, se houver k elementos após a remoção das duplicatas,
 * os primeiros k elementos de nums devem conter o resultado final.
 * Não importa o que você deixa além dos primeiros k elementos.<br><br>
 * <p>
 * Retorne k após colocar o resultado final nos primeiros k slots de nums.<br><br>
 * <p>
 * Não aloque espaço extra para outro array.
 * Você deve fazer isso modificando a matriz de entrada no local com memória extra O(1).<br><br>
 * <p>
 * Restrições:<br>
 * 0 <= nums.length <= 100<br>
 * 0 <= nums[i] <= 50<br>
 * 0 <= val <= 100<br>
 */
public class RemoveElement {

  public int removeElement(int[] nums, int val) {
    if (nums.length == 0) return 0;

    int indexStart = 0;
    int indexEnd = nums.length - 1;
    int answer = nums.length;

    while (indexStart <= indexEnd) {
      if (nums[indexStart] == val) {
        answer--;
        while (indexStart < indexEnd && nums[indexEnd] == val) {
          indexEnd--;
          answer--;

        }
        if (indexStart < indexEnd) {
          nums[indexStart] = nums[indexEnd--];
        }
      }

      indexStart++;
    }

    return answer;
  }

}

File: javascript/problems/RemoveElement.js
/**
 * Dado um array de inteiros nums e um inteiro val, remova todas as ocorrências de val em nums in-plice.
 * A ordem dos elementos pode ser alterada. Em seguida, retorne o número de elementos em nums que não são iguais a val.
 *
 * Considere o número de elementos em nums que não são iguais a val ser k, para ser aceito, você precisa fazer o seguinte:
 *
 * - Altere o array nums de modo que os primeiros k elementos de nums contenham os elementos que não são iguais a val.
 * Os elementos restantes de nums não são importantes, assim como o tamanho de nums.
 *
 * - Retorne k.
 *
 * Restrições:
 *
 * 0 <= nums.length <= 100
 *
 * 0 <= nums[i] <= 50
 *
 * 0 <= val <= 100
 *
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
var removeElement = function (nums, val) {
  let k = 0;
  for (const element of nums) {
    if (element !== val) {
      nums[k++] = element;
    }
  }
  return k;
};
